### PR TITLE
[doc] Add an ADR about hidden elements

### PR DESCRIPTION
**Pull request by theogiraudet**
# Pull request template

## General purpose
What is the main goal of this pull request?
- [ ] Bug fixes
- [ ] New features
- [x] Documentation
- [ ] Cleanup
- [ ] Tests
- [ ] Build / releng

## Project management
- [x] Has the pull request been added to the relevant project and milestone? (Only if you know that your work is part of a specific iteration such as the current one)
- [ ] Have the `priority:` and `pr:` labels been added to the pull request? (In case of doubt, start with the labels `priority: low` and `pr: to review later`)
- [ ] Have the relevant issues been added to the pull request?
- [ ] Have the relevant labels been added to the issues? (`area:`, `difficulty:`, `type:`)
- [ ] Have the relevant issues been added to the same project and milestone as the pull request?
- [ ] Has the `CHANGELOG.adoc` been updated to reference the relevant issues?
- [ ] Have the relevant API breaks been described in the `CHANGELOG.adoc`? (Including changes in the GraphQL API)
- [ ] In case of a change with a visual impact, are there any screenshots in the `CHANGELOG.adoc`? For example in `doc/screenshots/2022.5.0-my-new-feature.png`

## Architectural decision records (ADR)
- [x] Does the title of the commit contributing the ADR start with `[doc]`?
- [x] Are the ADRs mentioned in the relevant section of the  `CHANGELOG.adoc`?

## Dependencies
- [ ] Are the new / upgraded dependencies mentioned in the relevant section of the `CHANGELOG.adoc`?
- [ ] Are the new dependencies justified in the `CHANGELOG.adoc`?

## Frontend

This section is not relevant if your contribution does not come with changes to the frontend.

### General purpose
- [ ] Is the code properly tested? (Plain old JavaScript tests for business code and tests based on React Testing Library for the components)

### Typing
We need to improve the typing of our code, as such, we require every contribution to come with proper TypeScript typing for both changes contributing new files and those modifying existing files.
Please ensure that the following statements are true for each file created or modified (this may require you to improve code outside of your contribution).

- [ ] Variables have a proper type
- [ ] Functions’ arguments have a proper type
- [ ] Functions’ return type are specified
- Hooks are properly typed:
	- [ ] `useMutation<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useQuery<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useSubscription<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useMachine<CONTEXT_TYPE, EVENTS_TYPE>(…)`
	- [ ] `useState<STATE_TYPE>(…)`
- [ ] All components have a proper typing for their props
- [ ] No useless optional chaining with `?.` (if the GraphQL API specifies that a field cannot be `null`, do not treat it has potentially `null` for example)
- [ ] Nullable values have a proper type (for example `let diagram: Diagram | null = null;`)

## Backend

This section is not relevant if your contribution does not come with changes to the backend.

### General purpose
- [ ] Are all the event handlers tested?
- [ ] Are the event processor tested?
- [ ] Is the business code (services) tested?
- [ ] Are diagram layout changes tested?

### Architecture
- [ ] Are data structure classes properly separated from behavioral classes?
- [ ] Are all the relevant fields final?
- [ ] Is any data structure mutable? If so, please write a comment indicating why
- [ ] Are behavioral classes either stateless or side effect free?

### How to test this PR?
_Please describe here the various use cases to test this pull request_

- [ ] Has the Kiwi TCMS test suite been updated with tests for this contribution?

### CODE REVIEW

File: doc/adrs/070_add_the_possibility_to_hide_or_fade_a_graphic_element.adoc [hunk 1]
```diff
@@ -0,0 +1,67 @@
+= ADR-070 - Add the possibility to hide or fade a graphic element
```
Comment (sbegaudeau) [inline doc/adrs/070_add_the_possibility_to_hide_or_fade_a_graphic_element.adoc]: Using ASCIIDoc, you are supposed to use one line per sentence. It would make it easier to comment such PRs.

File: doc/adrs/070_add_the_possibility_to_hide_or_fade_a_graphic_element.adoc [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
...sibility` that is an enumeration defining `NORMAL`, `FADED` and `HIDDEN`. In this last case, some graphic information like the `position` and `size` doesn't have meaning. Indeed, if the auto-layout is enabled, ELK decides the value of these parameters but it ignores hidden elements (because of the `noLayout`). This can create runtime bugs in the existing code: if a function iterates over graphic elements and uses the position/size for computation, the result can be wrong due to hidden elements. To resolve that, we can make these graphic information attributes optional. The `Optional` is empty if the state is `HIDDEN` and wraps the information otherwise. All code sections requesting these values will throw a compilation error. This makes it easier to identify the code to be adapted and thus avoids hidden execution bugs as described. An other solution is to wrap all graphic information in a common class `GraphicNodeInformation` and `GraphicEdgeInformation` and change existing attributes to a single attribute of these type that is optional. This avoids testing the visibility (or to call `Optional#get`) multiple time if we need to access to several graphic information at the same time.
```
Comment (sbegaudeau) [inline doc/adrs/070_add_the_possibility_to_hide_or_fade_a_graphic_element.adoc]: - We could probably go with something like `ViewVisibility` instead of `GraphicalElementVisibility` since IIRC all the elements that can appear in a diagram are considered "views" (which explain the name of some variables like `sourceView`, `targetView`, `containerView` etc)
- Switching the types of `Node#size: Size`, `Node#position: Position` and `Edge#routingPoints: List<Position>` to `Node#size: Optional<Size>`, `Node#position: Optional<Position>` and `Edge#routingPoints: Optional<List<Position>>` would actually help save other issues. Right now, when we don't have any position for a node, it is initialized with `[-1, -1]` but the problem is that those coordinates are completely valid in a diagram. Using optional would make the lack of size/position/routingPoints explicit. It would move some existing bugs from being runtime bugs to compile time bugs which is good.

File: doc/adrs/070_add_the_possibility_to_hide_or_fade_a_graphic_element.adoc [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+==== Impacted code
```
Comment (sbegaudeau) [inline doc/adrs/070_add_the_possibility_to_hide_or_fade_a_graphic_element.adoc]: This is fine but you don't have to try to create an explicit list of all the pieces of code impacted. You could have said, the ELK layout, the incremental layout and this and that.

File: CHANGELOG.adoc [hunk 4]
```diff
@@ -5,6 +5,7 @@
 === Architectural decision records
 
 - [ADR-069] Add support for children layout strategy
+- [ADR-070] Add the possibility to hide or fade a graphic element
```
Comment (sbegaudeau) [inline CHANGELOG.adoc]: It could be better to by a tiny bit more precise since we have countless of "stuff" which can be considered as "graphical elements" with a title like "Add support to hide or fade diagram's views"

@pcdavid In Sirius Desktop all the graphical stuff in a diagram is called "view" right? like nodes, edges, decorators etc?

File: doc/adrs/070_add_the_possibility_to_hide_or_fade_a_graphic_element.adoc [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+=== Backend
+
+==== Auto-layout
+
+In Sirius Web, the auto-layout is managed by ELK. Sirius gives to ELK the graph and get from ELK a layouted graph. This is this layouted graph that is saved in database. When an element is hidden, we need to keep it in memory to allow the user to display it again. 
+
+The first solution is to totally remove a hidden element from the saved graph and to store it separately. But the element is just hidden, not removed and if a part of the application iterates over all elements of the graph, this part can not access to hidden elements. 
+
+The second solution is to persist these hidden elements after the auto-layout was applied. A way to do that is to indicate to ELK to not consider the element in the layout computation. For this, ELK proposes the https://www.eclipse.org/elk/reference/options/org-eclipse-elk-noLayout.html[`noLayout`] property applies to nodes, border nodes, edges and labels. With this solution hidden elements are still present in the diagram, to invisibilize them we can use CSS.
```
Comment (gcoutable) [inline doc/adrs/070_add_the_possibility_to_hide_or_fade_a_graphic_element.adoc]: Just in case: According to the documentation, the `noLayout` option does not say that such elements will be hidden. They will still be visible in the frontend but with unknown size and position.

Comment (sbegaudeau) [reply doc/adrs/070_add_the_possibility_to_hide_or_fade_a_graphic_element.adoc]: So we may need to ignore on the frontend those elements and the edges related to them

Comment (theogiraudet) [reply doc/adrs/070_add_the_possibility_to_hide_or_fade_a_graphic_element.adoc]: Maybe my text isn't very explicit: I propose to use CSS properties to hide elements such as edges or nodes. Indeed, the `noLayout` is just to avoid the "presence" of these hidden elements on the layout of the diagram.
An other solution is to not send these graphic elements to the front (better for performance).

File: CHANGELOG.adoc (hunk 4 above)
Comment (pcdavid) [reply CHANGELOG.adoc]: We use the term, and variables which expose these elements are indeed called `view*`, but strictly speaking there is no type called "View" (there is `DView` but it's the top-level element of an aird).
In Sirius Desktop we have:
* `DDiagramElement extends DRepresentationElement extends DSemanticDecorator` as the most general common type for all elements on a diagram.
* Then `DNode`, `DNodeContainer` and `DEdge` are subtypes of `DDiagramElement`

`DDiagramElement` is the one which has a `boolean isVisible()` attribute, but this one is "derived" (not in the common EMF sense, but via a pre-commit listener, `VisibilityUpdater`) from another attribute: `EList<GraphicalFilter> getGraphicalFilters()`.

Each diagram element can have one or more `GraphicalFilter` attached, which are used to compute the element's visibility (as in `isVisible`), and importantly, "adjacent" elements visiblity too. For example an edge is visible only if both its source *and* target are visible (see https://github.com/pcdavid/org.eclipse.sirius/blob/master/plugins/org.eclipse.sirius.diagram/src/org/eclipse/sirius/diagram/business/internal/helper/display/DisplayServiceImpl.java#L221); a label can be hidden either because it was hidden itself explicitly or because its parent node is invisible, etc.

`GraphicalFilter`s act like "modifiers". The main ones are:
* `HideFilter`: "basic" hide/reveal feature as expected in this ADR;
* `FoldingFilter`: can be applied to some edges (think an edge from a parent node to its children) to "fold" the edge and hide the whole sub-tree of elements "below" the edge;
* `CollapseFilter`: can be applied to a node or container to reduce its size and hide all its content.

This may seem over-complicated, but we need to store and remember each of these flags to restore the proper state when an element is revealed.

Basically an element can become "invisible" on the diagram for different, cumulative reasons, and removing *one* of these reasons is not enough to make it visible again, so a `boolean` is not enough.

File: doc/adrs/070_add_the_possibility_to_hide_or_fade_a_graphic_element.adoc [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+Concerning faded graphic elements, they is still visible (from the user point of view) so the auto-layout will be applied.
```
Comment (pcdavid) [inline doc/adrs/070_add_the_possibility_to_hide_or_fade_a_graphic_element.adoc]: "they is" => "they are"

**Comment by pcdavid** [comment]
As hinted in my comment about how it currently works in Sirius Desktop, I think the document is missing some explicit rules about how setting an element as FADED or HIDDEN can propagate (or not) to other adjacent elements (children, border nodes, labels, source/target elements for edges). In other words, the actual visibility and "display mode" (opaque or semi-transparent) of a given element can depend on the state of other related elements even if the element has no such attributes. Rules like:
* If either the source of target nodes of an edge are HIDDEN, the edge is considered implicitly HIDDEN;
* If the source of an edge is FADED, does it make the edge FADED? if the source *and* target are FADED, does this change anything? (it could be strange visually to have a fully visible edge between two semi-transparent elements, but that is debatable).
* If a container is HIDDEN, obviously all its children (recursively, incl. border nodes) should become invisible even if they have not been HIDDEN themselves explictly. Is it the same for FADED?
* etc.

**Comment by pcdavid** [comment]
Also, what kind of UI is planned to reveal the hidden/faded elements?
The document only mentions:

> the representation will propose a way to open a dialog displaying hidden elements with the possibility to make them visible again.

* But what is displayed exactly in this dialog? A flat list of the labels of the hidden/faded elements? If so, how are represented elements without labels? How to distinguish elements which have the same label?
* Is it the same UI/dialog to reveal/"unfade" elements?
* Can it be used to reveal/"unfade" multiple elements at the same time?
* Is it a modal dialog, which when open would hide most of the diagram?
* It could be a side-panel, which can be opened while keeping the diagram visible, so that the effect is immediatly visible.
* When revealing/unfading an element, is it selected afterwards?

File: CHANGELOG.adoc (hunk 4 above)
Comment (theogiraudet) [reply CHANGELOG.adoc]: Do I use `diagram elements` or `diagram's views`?
In this ADR, I propose to use an enum type to describe the state of the diagram element. This can be compared to a boolean because different possible states are incompatible with each other. Thus, an element is invisible (or faded) or not regardless of the reason. Do we have to implement a more flexible architecture to be able to handle the type of situation you mentioned in the future or should we keep it simple as I propose for the moment?

**Comment by mbats** [comment]
> Also, what kind of UI is planned to reveal the hidden/faded elements?

We prepared 3 different versions, the idea is to first implement the v1, then the v2 and later the v3. For 2022-11 we focus only on v1.

Please find attached videos demonstrating the expected behaviours:
* version 1 : 2 actions in the palette to hide/fade an element and 2 actions in the toolbar to show all the hidden/faded elements
https://user-images.githubusercontent.com/1185863/193817834-3cf17e7e-9c6e-417d-8e82-32257530a4dd.mp4
* version 2 : same as v1 but we are able to maintain how many elements are hidden/faded and show this information in the toolbar
https://user-images.githubusercontent.com/1185863/193818803-f1f7a55d-44f2-484a-b483-24df3916e0ed.mp4
* version 3 : a dialog is available in the toolbar to select elements which must be hidden/faded
https://user-images.githubusercontent.com/1185863/193819079-4be71176-f1f9-4fc4-8410-6caf91b4b7e1.mp4

File: CHANGELOG.adoc (hunk 4 above)
Comment (sbegaudeau) [reply CHANGELOG.adoc]: I think that there are two types of information we need to have for two different needs:

- On the frontend, we need to know if an element should be displayed on not. For that, we can use several solutions such as "not returning the element in question from the backend", "use a boolean to indicate that its hidden" or even "use an enum to indicate that its hidden"
- On the backend, we need to know why an element should be displayed. For that, we will need the details involved for example "some edge is hidden because it is between two nodes which are in two hidden containers because they are collapsed" (even if collapsing is out of scope for now)

So we may need to keep the modifiers placed on nodes and edges like in Sirius Desktop `Node#modifiers: List<Modifier>` with :

```
enum Modifier {
  HIDDEN,
  FADED,
  (maybe COLLAPSED one day)
}
```

But we may also need, for the frontend, to compute (and store?) the impact of those modifiers. This would prevent us from having to compute this impact every time we need to process a diagram. This information could be something like that: `Node#state: Modifier` and `Edge#state: Modifier`.

For example, let's consider the diagram in this screenshot: 

Capture d’écran 2022-10-14 à 15 26 13

If `Central_Unit` has for `modifiers` the values `[HIDDEN, FADED]`, then the following elements would be impacted with the following `state`:

- `DSP`: `HIDDEN`
- `Motion_Engine`: `HIDDEN`
- The fan with the label `100`: `HIDDEN`
- The edge with the label `10` between `DSP` and `Motion_Engine`: `HIDDEN`
- The edge with the label `6` between `Radar_Capture` and `Motion_Engine`: `HIDDEN`
- The edge with the label `10` between `GPU` and `Motion_Engine` should have the state `HIDDEN`

If on top of that, the `GPU` has for `modifiers` the values `[HIDDEN, FADED]`, then:

- The edge with the label `6` between `Back_Camera` and `GPU` should also have the state `HIDDEN`

As such, we need two pieces of information, the `modifiers` applied on the elements by the user and their impact on the `state` of a diagram element.

What do you think? @theogiraudet @pcdavid

Comment (theogiraudet) [reply CHANGELOG.adoc]: That sounds good to me, I don't see any case where this proposal doesn't work.

I see that you use a `List` to store the different modifiers. Is there a situation where we can have the same modifier more than once, e.g. `[HIDDEN, HIDDEN]` or is this an impossible case?

Comment (sbegaudeau) [reply CHANGELOG.adoc]: I don't think we have any reason to have a `List` instead of a `Set` indeed

File: doc/adrs/070_add_the_possibility_to_hide_or_fade_a_graphic_element.adoc (hunk 5 above)
Comment (sbegaudeau) [reply doc/adrs/070_add_the_possibility_to_hide_or_fade_a_graphic_element.adoc]: I don't think that we should use CSS for that, we can simply not tell Sprotty about the hidden elements. We should just ignore those hidden elements in the `convertDiagram` function in the frontend

**Review by sbegaudeau — APPROVED** [review]
